src/Nervos/Address.cpp

```diff
@@ -0,0 +1,160 @@
+        if (!Bech32::convertBits<5, 8, false>(decodedPayload, decodedData)) {
+            break;
+        }
+        if (decodedPayload.size() == 0) {
+            break;
+        }
+        addressType = (AddressType)decodedPayload[0];
+        switch (addressType) {
+        case AddressType::FullVersion: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32M) {
+                break;
+            }
+            if (decodedPayload.size() < 34) {
+                break;
+            }
+            codeHashIndex = -1;
+            codeHash = Data(decodedPayload.begin() + 1, decodedPayload.begin() + 33);
+            hashType = (HashType)decodedPayload[33];
+            args = Data(decodedPayload.begin() + 34, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        case AddressType::HashIdx: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32) {
+                break;
+            }
+            if (decodedPayload.size() != 22) {
+                break;
+            }
+            codeHashIndex = decodedPayload[1];
+            if (codeHashIndex != 0) {
+                break;
+            }
+            codeHash = CommonFunc::getSecp256k1CodeHash();
+            hashType = HashType::Type1;
+            args = Data(decodedPayload.begin() + 2, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        case AddressType::DataCodeHash:
+        case AddressType::TypeCodeHash: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32) {
+                break;
+            }
+            if (decodedPayload.size() < 33) {
+                break;
+            }
+            codeHashIndex = -1;
+            codeHash = Data(decodedPayload.begin() + 1, decodedPayload.begin() + 33);
+            hashType = addressType == AddressType::DataCodeHash ? HashType::Data0 : HashType::Type1;
+            args = Data(decodedPayload.begin() + 33, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        }
+    } while (false);
+
+    if (!isValid) {
+        throw std::invalid_argument("Invalid address string");
+    }
+}
+
+Address::Address(const PublicKey& publicKey)
+    : Address(publicKey, TW::p2pkhPrefix(TWCoinTypeNervos),
+              stringForHRP(TW::hrp(TWCoinTypeNervos))) {}
+
+Address::Address(const PublicKey& publicKey, byte p2pkh, const char* hrp) : hrp(hrp) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1) {
+        throw std::invalid_argument("Nervos::Address needs a SECP256k1 public key.");
+    }
+    addressType = AddressType::FullVersion;
+    codeHashIndex = -1;
+    codeHash = CommonFunc::getSecp256k1CodeHash();
+    hashType = HashType::Type1;
+    args = TW::Hash::blake2b(publicKey.bytes, 32, CommonFunc::getHashPersonalization());
+    args = Data(args.begin(), args.begin() + 20);
+}
+
+std::string Address::string() const {
+    auto data = Data();
+    data.push_back((byte)addressType);
```

Comment:
same remark for the c-style cast, also consider using `emplace_back` instead of `push_back` in order to construct the object directly in place

Reply:
Done